sdk/identity/azure-identity/inc/azure/identity/chained_token_credential.hpp

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Chained Token Credential.
+ */
+
+#pragma once
+
+#include <azure/core/credentials/credentials.hpp>
+
+#include <memory>
+#include <vector>
+
+namespace Azure { namespace Identity {
+  /**
+   * @brief Chained Token Credential provides a token credential implementation which chains
+   * multiple `%Core::Credentials::TokenCredential` implementations to be tried in order until one
+   * of the `%GetToken()` methods returns an access token.
+   *
+   */
+  class ChainedTokenCredential final : public Core::Credentials::TokenCredential {
+  public:
+    /**
+     * @brief A container type to store the ordered chain of credentials.
+     *
+     */
+    using Sources = std::vector<std::shared_ptr<Core::Credentials::TokenCredential>>;
```

Comment:
I see what you're saying, but it is a disadvantage of some development environments. You don't have to use it - in some cases you might be able to get away with just `ChainedTokenCredential cred({c1, c2});` initialization, but in general, I think it is better to write
```
Azure::Identity::ChainedTokenCredential cred(Azure::Identity::ChainedTokenCredential::Sources{ c1, c2 });
```
than
```
Azure::Identity::ChainedTokenCredential cred(std::vector<std::shared_ptr<Azure::Core::Credentials::TokenCredential>>{ c1, c2 });
```
especially that you can write
```
using Azure::Identity::ChainedTokenCredential;
ChainedTokenCredential cred(ChainedTokenCredential::Sources{ c1, c2 });
```
but there is not a single `using` that would shorten the second form.

Reply:
> but in general, I think it is better to write

The decision for adding a `using Type OtherLongerType` should be a customer's decision. That's what I think.
Customers can write the `using ...` and they will be familiar with the alias/type they create (instead of learning about the one the SDK library is providing).

I would like to hear from others in the team.  May be it is just me =P.  We should vote and make our guideline about it

Reply:
On a different note, why is this a vector of **shared pointers**?

Reply:
@vhvb1989, I get it as well, let's see what we gat from the Beta feedback.
@ahsonkhan, 1. Because in all other places when you as a user create a credential, you put it into a shared pointer, so it is "idiomatic". 2. It should be a smart pointer, because this credential needs the credentials to be alive while it exists. So, we basically left with a choice of `shared_ptr` (non-exclusive lifetime ownership), and `unique_ptr` (lifetime ownership). Technically, exclusive ownership is not necessary, shared is enough. This is also how it works on a conceptual level in .NET and Java. Plus it is tricky to pass a vector of unique_pointers by copy, as we do here Additionally, it would allow you to reuse the same credential on its own and as a part of Chained credential.

Reply:
To clarify a bit, ideally you'd want it to exclusive ownership.  However, the design of identity allows the credential to be shared, thus it ends up necessitating the shared_ptr.